Prevent broken concurrent schema read/writes

bereng opened the pull request.

bereng committed:
Prevent broken concurrent schema read/writes

bereng committed:
CIRLCE MID DO NOT MERGE

bereng commented:
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/477/workflows/0c21bd79-3217-4f08-b63d-ac258069ae0b)
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/477/workflows/19b1d3e7-6881-4e88-bd10-7df66d1df19b)

Failures look unrelated, timeouts or known

maedhroz commented on src/java/org/apache/cassandra/schema/SchemaKeyspaceTables.java:
```diff
@@ -0,0 +1,52 @@
+
```
nit: random newline?

maedhroz commented on src/java/org/apache/cassandra/schema/SchemaKeyspaceTables.java:
```diff
@@ -0,0 +1,52 @@
+     * See CASSANDRA-12213 for more details.
+     */
+    public static final ImmutableList<String> ALL =
+        ImmutableList.of(COLUMNS, DROPPED_COLUMNS, TRIGGERS, TYPES, FUNCTIONS, AGGREGATES, INDEXES, TABLES, VIEWS, KEYSPACES);
```
nit: seems like this has a newline for every table name in 4.0...don't really care which format we go with, but probably good to be consistent

bereng replied:
It has to be the other one bc of long lines and the formatting rules. Solved.

maedhroz commented on src/java/org/apache/cassandra/schema/SchemaKeyspace.java:
```diff
@@ -70,41 +75,14 @@ private SchemaKeyspace()
 
     private static final TableMetadata Keyspaces =
-        parse(KEYSPACES,
+        parse(SchemaKeyspaceTables.KEYSPACES,
```
@bereng Seems like in the 4.0 PR, we added a static import for the constants in `SchemaKeyspaceTables` to cut down on the size of the diff. Don't care a ton about which way we go, but let's be consistent.

maedhroz reviewed (changes requested):
Left some minor comments, but 2 things I think we have to fix for sure. Looks like we're missing the changes from the 4.0 patch for `MigrationManager` and `SchemaPullVerbHandler`.

bereng commented:
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/482/workflows/f491df19-9629-4198-a75f-e015b30962b7)
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/482/workflows/e837aaf2-a3c7-4a2f-8358-7e8ceb891906)

maedhroz commented on test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:
```diff
@@ -63,16 +69,71 @@ public static void defineSchema() throws ConfigurationException
+
+    /** See CASSANDRA-16856/16996. Make sure schema pulls are synchronized to prevent concurrent schema pull/writes */
+    @Test
+    public void testNoVisiblePartialSchemaUpdates() throws Exception
```
@bereng So we still need the Byteman rule here, because otherwise there's nothing stopping the table creation from finishing completely before we call `Schema.instance.schemaKeyspaceAsMutations()`. In other words, we still need some arbitrarily large wait on `CFS#apply()` to make sure the second thread will see the mutation partially applied.

bereng commented:
Rebased, squashed
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/483/workflows/c03b2d1d-ed45-4cd8-9f2a-5eb5cf99294c)
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/483/workflows/0baa4102-eef0-42c8-aa7c-ea830efae4b5)
